colossalai/nn/optimizer/distributed_lamb.py

```diff
@@ -0,0 +1,177 @@
+            self.dp_size = dist.get_world_size(dp_group)
+
+        self.shard_to_param = shard_to_param if shard_to_param is not None else {}
+        self.is_zero = is_zero
+        self.is_dist = {}
+        # Cache parameter layout
+        for group in self.param_groups:
+            for p in group["params"]:
+                self.is_dist[p] = (
+                    is_distributed_tensor(p)
+                    if self.dp_size <= 1
+                    else is_distributed_tensor(self.shard_to_param.get(id(p), None))
+                )
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step.
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad.data
+                if grad.is_sparse:
+                    raise RuntimeError("Lamb does not support sparse gradients, consider SparseAdam instad.")
+
+                state = self.state[p]
+                # State initialization
+                if len(state) == 0:
+                    state["step"] = 0
+                    # Exponential moving average of gradient values
+                    state["exp_avg"] = torch.zeros_like(p.data)
+                    # Exponential moving average of squared gradient values
+                    state["exp_avg_sq"] = torch.zeros_like(p.data)
+
+                exp_avg, exp_avg_sq = state["exp_avg"], state["exp_avg_sq"]
+                beta1, beta2 = group["betas"]
+
+                state["step"] += 1
+
+                # Decay the first and second moment running average coefficient
+                # m_t
+                exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+                # v_t
+                exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+                scaled_lr = group["lr"]
+                if group["bias_correction"]:
+                    bias_correction1 = 1 - beta1 ** state["step"]
+                    bias_correction2 = 1 - beta2 ** state["step"]
+                    # Apply debiasing to lr to avoid broadcast
+                    scaled_lr *= (bias_correction2**0.5) / bias_correction1
+                    # exp_avg.div_(bias_correction1)
+                    # exp_avg_sq.div_(bias_correction2)
+
+                update = exp_avg / exp_avg_sq.sqrt().add(group["eps"])
+                if group["weight_decay"] != 0:
+                    update.add_(p.data, alpha=group["weight_decay"])
+
+                # Compute global layer-wise trust ratio
+                if self.is_dist[p]:
+                    p_local = p
+                    g_sum = (update**2).sum()
+                    if self.dp_size > 1 and self.is_zero:
```

Comment:
The logic is wrong when TP is totally disabled. E.g. we only use zero and there is no dtensor, but grads also need to be reduced.